networkx/algorithms/approximation/steinertree.py

```diff
@@ -46,22 +57,172 @@ def metric_closure(G, weight="weight"):
     return M
 
 
+def mehlhorn_steiner_tree(G, S, weight):
+    paths = nx.multi_source_dijkstra_path(G, S)
+
+    d_1 = {}
+    s = {}
+    for v in G.nodes():
+        s[v] = paths[v][0]
+        d_1[(v, s[v])] = len(paths[v]) - 1
+
+    new_edges = {}
+    for e in G.edges(data=True):
+        u, v, data = e
+        weight_here = d_1[(u, s[u])] + data.get(weight, 1) + d_1[(v, s[v])]
+        key = tuple(sorted([s[u], s[v]]))
+        new_edges[key] = min(new_edges.get(key, float("inf")), weight_here)
+
+    G_1_prime = nx.Graph()
+    for (n1, n2), edge_weight in new_edges.items():
+        G_1_prime.add_edge(n1, n2, weight=edge_weight)
+
+    G_2 = nx.minimum_spanning_edges(G_1_prime, data=True)
+
+    G_3 = nx.Graph()
+    for u, v, d in G_2:
+        path = nx.shortest_path(G, u, v, weight)
+        for n1, n2 in pairwise(path):
+            G_3.add_edge(n1, n2)
+
+    G_4 = nx.minimum_spanning_edges(G_3, data=False)
+    return G_4
+
+
+def kou_steiner_tree(G, terminal_nodes, weight):
+    # H is the subgraph induced by terminal_nodes in the metric closure M of G.
+    M = metric_closure(G, weight=weight)
+    H = M.subgraph(terminal_nodes)
+    # Use the 'distance' attribute of each edge provided by M.
+    mst_edges = nx.minimum_spanning_edges(H, weight="distance", data=True)
+    # Create an iterator over each edge in each shortest path; repeats are okay
+    edges = chain.from_iterable(pairwise(d["path"]) for u, v, d in mst_edges)
+    return edges
+
+
+def wu_steiner_tree(G, terminal_nodes, weight):
+    # Step 1
+    source = {q: q for q in terminal_nodes}
+    length = {q: 0 for q in terminal_nodes}
+    pred = {}
+
+    # Step 2
+    both_in_terminal_nodes = set()
+    priority_queue = []
+    for q in terminal_nodes:
+        for r in G.neighbors(q):
+            if r in terminal_nodes:
+                both_in_terminal_nodes.add((min(r, q), max(r, q)))
```

Comment:
Avoid sorting/comparing the nodes here by using a frozenset... The two directions will combine into one in the set both_in_terminal_nodes. And it will still work to iterate `for q, r in both...`
```both_in_terminal_nodes.add(frozenset(q, r))```